### ISSUE TITLE
Collscan on Browsing a Collection

### ISSUE THREAD

**Issue by Xeroxxx** [mongo-express/mongo-express#1518]
I've the problem that when I open a large collection (~10.000.000 Obj, 22kb Avg, 100GB).

I takes ages to load and I can see a COLLSCAN of _id in the logs.

```
mongo-1          | {"t":{"$date":"2024-04-20T23:47:40.771+00:00"},"s":"I",  "c":"COMMAND",  "id":51803,   "ctx":"conn26","msg":"Slow query","attr":{"type":"command","ns":"xxx.xxx",
"command":{"aggregate":"xxx","pipeline":[{"$match":{}},{"$group":{"_id":1,"n":{"$sum":1}}}],"cursor":{},"lsid":{"id":{"$uuid":"2b51c014-62f7-49d0-8279-979a8dfa868d"}},"$db":"poedata"},
"planSummary":"COLLSCAN","planningTimeMicros":166,"keysExamined":0,"docsExamined":9343804,"fromPlanCache":true,"cursorExhausted":true,"numYields":47869,"nreturned":1,"queryHash":"366C3F12","planCacheKey":"58CF6002","queryFramework":"sbe","reslen":128,"locks":{"FeatureCompatibilityVersion":{"acquireCount":{"r":47870}},"Global":{"acquireCount":{"r":47870}}},"storage":{"data":{"bytesRead":215941107656,"timeReadingMicros":641289510},"timeWaitingMicros":{"cache":588}},"cpuNanos":351818832796,"remote":"172.18.0.3:46260","protocol":"op_msg","durationMillis":861631}}
```

The extracted query of mongo express is an aggregate on:
```
{"$match":{}},{"$group":{"_id":1,"n":{"$sum":1}}}
```

Replacing the match with for ex sort uses the _id index and then its fast.

Can anyone explain why the Index is not used for browsing a collection and why the index is not used for _id?

**Comment by xTudoS** [comment]
Are you using the master version or 1.0.2 ?

**Comment by Xeroxxx** [comment]
> Are you using the master version or 1.0.2 ?

I'm using latest docker image: https://hub.docker.com/_/mongo-express

Its tagged as 1.0.2-20

**Comment by xTudoS** [comment]
> > Are you using the master version or 1.0.2 ?
> 
> I'm using latest docker image: https://hub.docker.com/_/mongo-express
> 
> Its tagged as 1.0.2-20

try to build and use your own mongo-express image from master
use this repo to it: https://github.com/mongo-express/mongo-express-docker

official docker image out-of-date

**Comment by rtritto** [comment]
@Xeroxxx please can you test with the branch https://github.com/mongo-express/mongo-express/tree/improve-query-0?

**Comment by Xeroxxx** [comment]
> @Xeroxxx please can you test with the branch https://github.com/mongo-express/mongo-express/tree/improve-query-0?

I will check that in the upcoming days! Thank you!

**Comment by Xeroxxx** [comment]
@rtritto 
So the first collection that was loading before in like 10s throws this error and the second big collection throws this as well:
`PlanExecutor error during aggregation :: caused by :: document constructed by $facet is 104859566 bytes, which exceeds the limit of 104857600 bytes`

**Comment by Xeroxxx** [comment]
Found the issues I first noticed the issue after upgrading to 6.0 and using Alpha 4.

https://github.com/mongo-express/mongo-express/issues/887#issuecomment-1297683795

**Comment by rtritto** [comment]
@Xeroxxx I replaced `$facet` with `$group`, please can you try?

**Comment by Xeroxxx** [comment]
So the one taking 10 seconds before, loads in less than 10 seconds. The big one with rougly 10.000.000 Elements:
`PlanExecutor error during aggregation :: caused by :: Used too much memory for a single array. Memory limit: 104857600 bytes. The array contains 2529 elements and is of size 104851560 bytes. The element being added has size 101356 bytes.`

**Comment by rtritto** [comment]
@Xeroxxx thanks for the availability, but, at the moment, the session isn't necessary.
We need an idea for the query, I opened a question: https://stackoverflow.com/questions/78583283

**Comment by rtritto** [comment]
@Xeroxxx I restored `find` and used `count` instead of `countDocuments`
Please can you try?

**Comment by Xeroxxx** [comment]
> @Xeroxxx I restored `find` and used `count` (deprecated) instead of `countDocuments`. Please can you try?

This does the job. Everythin loads within seconds! Great job!

**Comment by rtritto** [comment]
https://www.mongodb.com/docs/manual/reference/method/db.collection.count
> NOTE: MongoDB drivers compatible with the 4.0 features deprecate their respective cursor and collection count() APIs in favor of new APIs for countDocuments() and estimatedDocumentCount(). For the specific API names for a given driver, see the driver documentation.

From Doc of `count` method in _mongodb.d.ts_:
> An estimated count of matching documents in the db to a filter.
 **NOTE:** This method has been deprecated, since it does not provide an accurate count of the documents
 in a collection. To obtain an accurate count of documents in the collection, use {@link Collection#countDocuments| countDocuments}.
 To obtain an estimated count of all documents in the collection, use {@link Collection#estimatedDocumentCount| estimatedDocumentCount}.
 @deprecated use {@link Collection#countDocuments| countDocuments} or {@link Collection#estimatedDocumentCount| estimatedDocumentCount} instead

What should we do?

**Comment by rtritto** [comment]
@Xeroxxx please can you test the `aggregate` query?

Steps:
- Click `Advanced`
- Enable `Aggregate query`
- Write `[]` (empty list) in Query
- Click `Find`

I don't know if `aggregate` has a limit size and it will break; to fix it, the option `allowDiskUse=true` is needed.

Note: I can replace the _Promise.all_ of `aggregate` query and `aggregate` count with single `aggregate` query ($facet + $slice + $size) that is more performant but have the 100 MB in output $facet and 16MB in output of query.

**Comment by Xeroxxx** [comment]
> * Click `Advanced`
> * Enable `Aggregate query`
> * Write `[]` (empty list) in Query
> * Click `Find`

This query is loading forever again and never finishes.

**Comment by rtritto** [comment]
@Xeroxxx I should fix the `aggregate`, please can you try?

**Comment by Xeroxxx** [comment]
@rtritto The steps you mentioned above are not working fine (Advanced Tab)

**Comment by rtritto** [comment]
Closed by #1563
